### PR TITLE
Stardust ledger

### PR DESCRIPTION
**Pull request by thibault-martinez**
## Change checklist

Tick the boxes that are relevant to your changes, and delete any items that are not.

- [x] I have followed the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that new and existing unit tests pass locally with my changes
- [x] I have updated the CHANGELOG.md, if my changes are significant enough

### CODE REVIEW

File: bee-tangle/src/conflict.rs [hunk 1]
```diff
@@ -28,10 +28,22 @@ pub enum ConflictReason {
...
     /// The referenced Utxo cannot be found.
     InputUtxoNotFound = 3,
-    /// The sum of the inputs and output values does not match.
-    InputOutputSumMismatch = 4,
+    /// The created amount does not match the consumed amount.
+    CreatedConsumedAmountMismatch = 4,
     /// The unlock block signature is invalid.
     InvalidSignature = 5,
+    /// The dust allowance for the address is invalid.
+    InvalidDustAllowance = 6,
```
Comment (Thoralf-M) [inline bee-tangle/src/conflict.rs]: StorageDeposit?

Comment (thibault-martinez) [reply bee-tangle/src/conflict.rs]: Yeah definitely, still a draft, don't read!

File: bee-api/bee-rest-api/src/types/dtos.rs [hunk 2]
```diff
@@ -302,22 +295,14 @@ impl TryFrom<&InputDto> for Input {
...
-                    .into(),
-            )),
+            InputDto::Utxo(u) => Ok(Input::Utxo(u.try_into()?)),
+            InputDto::Treasury(t) => Ok(Input::Treasury(t.try_into()?)),
         }
     }
 }
 
+//.map_err(|_| Error::InvalidField("transactionId"))?,
+
```
Comment (grtlr) [inline bee-api/bee-rest-api/src/types/dtos.rs]: Maybe remove?

File: bee-message/src/unlock_block/mod.rs [hunk 3]
```diff
@@ -119,7 +119,7 @@ fn verify_unlock_blocks<const VERIFY: bool>(unlock_blocks: &[UnlockBlock]) -> Re
             UnlockBlock::Reference(reference) => {
                 if index == 0
                     || reference.index() >= index as u16
```
Comment (grtlr) [inline bee-message/src/unlock_block/mod.rs]: Maybe we don't need the cast here because index is already declared as `(u16..)`.

```suggestion
                    || reference.index() >= index
```

File: bee-message/src/address/ed25519.rs [hunk 4]
```diff
@@ -21,34 +18,13 @@ impl Ed25519Address {
     /// The [`Address`](crate::address::Address) kind of an [`Ed25519Signature`].
```
Comment (grtlr) [inline bee-message/src/address/ed25519.rs]: Do we still need the `#[allow(clippy::...)]`?

File: bee-message/tests/treasury_transaction_payload.rs [hunk 5]
```diff
@@ -25,49 +16,19 @@ fn kind() {
 
 #[test]
 fn new_valid() {
-    let input = Input::from(TreasuryInput::from_str(MESSAGE_ID).unwrap());
-    let output = Output::from(TreasuryOutput::new(1_000).unwrap());
-    let transaction = TreasuryTransactionPayload::new(input.clone(), output.clone()).unwrap();
+    let input = TreasuryInput::from_str(MESSAGE_ID).unwrap();
+    let output = TreasuryOutput::new(1_000).unwrap();
+    let transaction = TreasuryTransactionPayload::new(input, output.clone()).unwrap();
```
Comment (grtlr) [inline bee-message/tests/treasury_transaction_payload.rs:21]: Can the `clone` be removed here, too?
```suggestion
    let transaction = TreasuryTransactionPayload::new(input, output).unwrap();
```

File: bee-api/bee-rest-api/src/types/dtos.rs (hunk 2 above)
Comment (Adam-Gleave) [inline bee-api/bee-rest-api/src/types/dtos.rs]: Should this be removed?

File: bee-ledger/src/workers/consensus/white_flag.rs [hunk 6]
```diff
@@ -68,70 +58,155 @@ fn apply_regular_essence<B: StorageBackend>(
...
+                output.unlock_conditions(),
+            ),
+            Output::Nft(output) => (
+                output.unlock(output_id, unlock_block, &inputs, &mut context),
+                output.amount(),
+                output.native_tokens(),
+                output.unlock_conditions(),
+            ),
+            _ => return Err(Error::UnsupportedOutputKind(consumed_output.kind())),
+        };
```
Comment (Thoralf-M) [inline bee-ledger/src/workers/consensus/white_flag.rs:116]: This should rather be something like
```Rust
        if consumed_output.kind() != ValidOutputsRange {
            return Err(Error::UnsupportedOutputKind(consumed_output.kind()));
        }
        let (conflict, amount, consumed_native_tokens, unlock_conditions) = (
            consumed_output.unlock(output_id, unlock_block, &inputs, &mut context),
            consumed_output.amount(),
            consumed_output.native_tokens(),
            consumed_output.unlock_conditions(),
        );

Comment (thibault-martinez) [reply bee-ledger/src/workers/consensus/white_flag.rs:116]: But then you match 4 times for each output :(

File: bee-message/src/semantic.rs [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+    output::{inputs_commitment, ChainId, Output, OutputId, TokenId},
+    payload::transaction::{RegularTransactionEssence, TransactionEssence, TransactionId},
+    unlock_block::UnlockBlocks,
+};
+
+use hashbrown::{HashMap, HashSet};
+use primitive_types::U256;
+
+// Copyright 2020-2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
```
Comment (Thoralf-M) [inline bee-message/src/semantic.rs]:
```suggestion
```

Comment (thibault-martinez) [reply bee-message/src/semantic.rs]: Whoops

File: bee-message/src/semantic.rs [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+    pub unlocked_addresses: HashSet<Address>,
+}
+
+impl<'a> ValidationContext<'a> {
+    ///
+    pub fn new(
+        transaction_id: &TransactionId,
+        essence: &'a RegularTransactionEssence,
+        inputs: impl Iterator<Item = (&'a OutputId, &'a Output)> + Clone,
+        unlock_blocks: &'a UnlockBlocks,
```
Comment (Thoralf-M) [inline bee-message/src/semantic.rs:140]: Wouldn't it be cleaner to just accept a TransactionPayload here (of course the inputs need to be separated)?

Comment (thibault-martinez) [reply bee-message/src/semantic.rs:140]: I asked myself the same at some point. But the validation context is kind of tied to the essence type and not the transaction type. I need to think about it a bit more.

File: bee-ledger/src/workers/consensus/white_flag.rs [hunk 9]
```diff
@@ -68,70 +58,155 @@ fn apply_regular_essence<B: StorageBackend>(
...
+        context.input_amount = context
+            .input_amount
+            .checked_add(amount)
+            .ok_or(Error::ConsumedAmountOverflow)?;
+        for native_token in consumed_native_tokens.iter() {
+            let native_token_amount = context.input_native_tokens.entry(*native_token.token_id()).or_default();
+            *native_token_amount = native_token_amount
+                .checked_add(*native_token.amount())
+                .ok_or(Error::ConsumedNativeTokensAmountOverflow)?;
+        }
```
Comment (Adam-Gleave) [inline bee-ledger/src/workers/consensus/white_flag.rs]: I would add some newlines here for clarity:
```suggestion
        context.input_amount = context
            .input_amount
            .checked_add(amount)
            .ok_or(Error::ConsumedAmountOverflow)?;

        for native_token in consumed_native_tokens.iter() {
            let native_token_amount = context.input_native_tokens.entry(*native_token.token_id()).or_default();

            *native_token_amount = native_token_amount
                .checked_add(*native_token.amount())
                .ok_or(Error::ConsumedNativeTokensAmountOverflow)?;
        }
```

File: bee-ledger/src/workers/consensus/white_flag.rs (hunk 6 above)
Comment (Adam-Gleave) [inline bee-ledger/src/workers/consensus/white_flag.rs:116]: Maybe some shared method here could help with this repetition? I'm not sure whether its worth the effort or the indirection, but this code feels a little like it could do with some clean-up.

File: bee-ledger/src/workers/consensus/white_flag.rs [hunk 10]
```diff
@@ -68,70 +58,155 @@ fn apply_regular_essence<B: StorageBackend>(
...
+            .ok_or(Error::CreatedAmountOverflow)?;
+        for native_token in created_native_tokens.iter() {
+            let native_token_amount = *context
+                .output_native_tokens
+                .entry(*native_token.token_id())
+                .or_default();
+            native_token_amount
+                .checked_sub(*native_token.amount())
+                .ok_or(Error::CreatedNativeTokensAmountOverflow)?;
+        }
```
Comment (Adam-Gleave) [inline bee-ledger/src/workers/consensus/white_flag.rs]: Would add some newlines here again just to break up the block of text a bit.

File: bee-message/src/address/mod.rs [hunk 11]
```diff
@@ -67,4 +73,61 @@ impl Address {
...
+                // TODO actually check that it was unlocked by the same signature.
+                if !context.unlocked_addresses.contains(self) {
+                    return Err(ConflictReason::InvalidSignature);
+                }
+            }
+            (Address::Alias(alias_address), UnlockBlock::Alias(unlock_block)) => {
+                // SAFETY: indexing is fine as it is already syntactically verified that indexes reference below.
+                if let (output_id, Output::Alias(alias_output)) = inputs[unlock_block.index() as usize] {
+                    if &alias_output.alias_id().or_from_output_id(output_id) != alias_address.alias_id()
+                        || !context.unlocked_addresses.contains(self)
```
Comment (Adam-Gleave) [inline bee-message/src/address/mod.rs]: I don't really like multiline conditionals (there's a few in other files too), I think it's usually clearer if they're broken down a bit beforehand. At least for conditions that aren't straightforward. That could just be a personal nitpick, though.

File: bee-message/src/payload/receipt/mod.rs [hunk 12]
```diff
@@ -39,51 +44,55 @@ pub struct ReceiptPayload {
...
-    /// The `TreasuryTransaction` used to fund the funds of a `ReceiptPayload`.
-    pub fn transaction(&self) -> &Payload {
-        &self.transaction
+    /// The [`TreasuryTransactionPayload`](crate::payload::treasury_transaction::TreasuryTransactionPayload) used to
+    /// fund the funds of a [`ReceiptPayload`].
+    pub fn transaction(&self) -> &TreasuryTransactionPayload {
+        if let Payload::TreasuryTransaction(ref transaction) = self.transaction {
+            transaction
+        } else {
+            unreachable!()
```
Comment (Adam-Gleave) [inline bee-message/src/payload/receipt/mod.rs:92]: I would put a comment here just explaining (even though it's obvious, I feel like stuff that can panic should always have a bit of a note in case APIs change etc).

File: bee-message/src/payload/treasury_transaction/mod.rs [hunk 13]
```diff
@@ -16,25 +20,33 @@ pub struct TreasuryTransactionPayload {
...
 
-    /// Returns the input of a `TreasuryTransaction`.
-    pub fn input(&self) -> &Input {
-        &self.input
+    /// Returns the input of a [`TreasuryTransactionPayload`].
+    pub fn input(&self) -> &TreasuryInput {
+        if let Input::Treasury(ref input) = self.input {
+            input
+        } else {
+            unreachable!()
```
Comment (Adam-Gleave) [inline bee-message/src/payload/treasury_transaction/mod.rs:40]: Same as above.

File: bee-message/src/payload/treasury_transaction/mod.rs [hunk 14]
```diff
@@ -16,25 +20,33 @@ pub struct TreasuryTransactionPayload {
...
 
-    /// Returns the output of a `TreasuryTransaction`.
-    pub fn output(&self) -> &Output {
-        &self.output
+    /// Returns the output of a [`TreasuryTransactionPayload`].
+    pub fn output(&self) -> &TreasuryOutput {
+        if let Output::Treasury(ref output) = self.output {
+            output
+        } else {
+            unreachable!()
```
Comment (Adam-Gleave) [inline bee-message/src/payload/treasury_transaction/mod.rs:50]: Same as above :trollface:

File: bee-ledger/src/workers/consensus/white_flag.rs (hunk 6 above)
Comment (thibault-martinez) [reply bee-ledger/src/workers/consensus/white_flag.rs:116]: We actually have these methods on the `Output` type itself. But they internally match so it would mean matching 4 times for each output...

Comment (Adam-Gleave) [reply bee-ledger/src/workers/consensus/white_flag.rs:116]: Ah, that's a pain!

File: bee-ledger/src/workers/consensus/white_flag.rs [hunk 15]
```diff
@@ -68,54 +58,190 @@ fn apply_regular_essence<B: StorageBackend>(
...
-            }
-            Output::Alias(_) => {
-                // TODO
-            }
-            Output::Foundry(_) => {
-                // TODO
+    let inputs: Vec<(OutputId, &Output)> = consumed_outputs
+        .iter()
+        .map(|(output_id, created_output)| (*output_id, created_output.inner()))
+        .collect();
```
Comment (grtlr) [inline bee-ledger/src/workers/consensus/white_flag.rs:72]: Is the collect necessary here, or could we just use iterators for both?

File: bee-ledger/src/workers/error.rs [hunk 16]
```diff
@@ -60,11 +60,17 @@ pub enum Error {
...
     InvalidLedgerUnspentState(u64),
     /// Consumed amount overflow.
-    #[error("consumed amount overflow: {0}")]
-    ConsumedAmountOverflow(u128),
+    #[error("consumed amount overflow")]
+    ConsumedAmountOverflow,
     /// Created amount overflow.
-    #[error("created amount overflow: {0}")]
-    CreatedAmountOverflow(u128),
+    #[error("created amount overflow.")]
```
Comment (grtlr) [inline bee-ledger/src/workers/error.rs]:
```suggestion
    #[error("created amount overflow")]
```
The same goes for the others as well.

File: bee-message/src/output/unlock_condition/immutable_alias_address.rs [hunk 17]
```diff
@@ -36,12 +36,8 @@ impl ImmutableAliasAddressUnlockCondition {
 }
 
 fn verify_alias_address<const VERIFY: bool>(address: &Address) -> Result<(), Error> {
-    if VERIFY {
-        if let Address::Alias(_) = address {
-            Ok(())
-        } else {
-            Err(Error::InvalidAddressKind(address.kind()))
-        }
+    if VERIFY && !matches!(address, Address::Alias(_)) {
```
Comment (grtlr) [inline bee-message/src/output/unlock_condition/immutable_alias_address.rs]: Maybe we should have `is_alias`, `is_nft`, and `is_ed25519` methods in `Address`.

**Review by samuel-rufi — DISMISSED** [review]

File: bee-message/src/address/mod.rs [hunk 18]
```diff
@@ -67,4 +73,63 @@ impl Address {
         // SAFETY: encoding can't fail as `self` has already been validated and built.
         bech32::encode(hrp.as_ref(), self.pack_to_vec().to_base32(), Variant::Bech32).unwrap()
     }
+
+    ///
```
Comment (Alex6323) [inline bee-message/src/address/mod.rs:92]: docs

File: bee-message/src/output/alias_id.rs [hunk 19]
```diff
@@ -8,8 +8,15 @@ impl_id!(pub AliasId, 20, "TODO.");
...
-    fn from(output_id: &OutputId) -> Self {
+impl From<OutputId> for AliasId {
+    fn from(output_id: OutputId) -> Self {
         Self::from(output_id.hash())
     }
 }
+
+impl AliasId {
+    ///
+    pub fn or_from_output_id(self, output_id: OutputId) -> AliasId {
```
Comment (Alex6323) [inline bee-message/src/output/alias_id.rs]: missing docs;
I'ld use `Self` as return type instead of the actual type name;
The API is confusing me: I already have an instance of 'AliasId', and call a constructor like 'from' method on it.

File: bee-message/src/lib.rs [hunk 20]
```diff
@@ -31,6 +31,8 @@ pub mod output;
 pub mod parent;
 /// A module that provides types and syntactic validations of payloads.
 pub mod payload;
+/// A module that provides types and rules for semantic validation.
+pub mod semantic;
```
Comment (Alex6323) [inline bee-message/src/lib.rs:35]: maybe module should be named after the "thing" that is being validated like with the "payload" module.
If that for some reason is not possible, then I'ld probably still name the module "semantic_validation" for the time being despite it being longer.

File: bee-message/src/output/alias.rs [hunk 21]
```diff
@@ -296,6 +298,141 @@ impl AliasOutput {
             .map(|unlock_condition| unlock_condition.address())
             .unwrap()
     }
+
+    ///
```
Comment (Alex6323) [inline bee-message/src/output/alias.rs:302]: docs

File: bee-message/src/output/alias.rs [hunk 22]
```diff
@@ -296,6 +298,141 @@ impl AliasOutput {
...
+    #[inline(always)]
+    pub fn chain_id(&self) -> ChainId {
+        ChainId::Alias(self.alias_id)
+    }
+
+    ///
```
Comment (Alex6323) [inline bee-message/src/output/alias.rs:308]: docs

File: bee-message/tip.md [hunk 23]
```diff
@@ -0,0 +1,73 @@
+ce4feebb00f0dd7968b58334465651b95bd21722
```
Comment (grtlr) [inline bee-message/tip.md]: Should this file be committed as well?

File: bee-message/src/output/alias.rs [hunk 24]
```diff
@@ -296,6 +298,141 @@ impl AliasOutput {
...
+    pub fn unlock(
+        &self,
+        output_id: &OutputId,
+        unlock_block: &UnlockBlock,
+        inputs: &[(OutputId, &Output)],
+        context: &mut ValidationContext,
+    ) -> Result<(), ConflictReason> {
+        let alias_id = if self.alias_id().is_null() {
+            AliasId::new(output_id.hash())
```
Comment (Alex6323) [inline bee-message/src/output/alias.rs]: wasn't there a `From<OutputId>` impl? Is this `new` ctor really necessary?

File: bee-message/src/output/alias.rs [hunk 25]
```diff
@@ -296,6 +298,141 @@ impl AliasOutput {
...
+
+        let locked_address = match next_state {
+            Some(Output::Alias(next_state)) => {
+                if self.state_index() == next_state.state_index() {
+                    self.governor_address()
+                } else {
+                    self.state_controller_address()
+                }
+            }
+            Some(_) => unreachable!(),
```
Comment (Alex6323) [inline bee-message/src/output/alias.rs]: can you document why that is unreachable?

File: bee-message/src/output/foundry.rs [hunk 26]
```diff
@@ -303,34 +305,87 @@ impl FoundryOutput {
...
+
+impl StateTransition for FoundryOutput {
+    fn creation(next_state: &Self, context: &ValidationContext) -> Result<(), StateTransitionError> {
+        let alias_chain_id = ChainId::from(*next_state.alias_address().alias_id());
+
+        if let (Some(_input_alias), Some(_output_alias)) = (
+            context.input_chains.get(&alias_chain_id),
+            context.output_chains.get(&alias_chain_id),
+        ) {
+            // TODO check serial
```
Comment (grtlr) [inline bee-message/src/output/foundry.rs:362]: Do you want to do this in a separate PR?

File: bee-message/src/output/alias.rs [hunk 27]
```diff
@@ -296,6 +298,141 @@ impl AliasOutput {
...
+
+        Ok(())
+    }
+
+    fn transition(
+        current_state: &Self,
+        next_state: &Self,
+        context: &ValidationContext,
+    ) -> Result<(), StateTransitionError> {
+        if current_state.immutable_feature_blocks != next_state.immutable_feature_blocks {
```
Comment (Alex6323) [inline bee-message/src/output/alias.rs:374]: wouldn't `validate_state_transition` be a better describing name for what this method actually does?

Comment (Alex6323) [reply bee-message/src/output/alias.rs:374]: also I think because this is probably code that should under no circumstances be wrong, it could either be a bit more described what needs to be checked or at least referred to the corresponding TIP.

File: bee-message/src/output/basic.rs [hunk 28]
```diff
@@ -172,6 +174,19 @@ impl BasicOutput {
             .map(|unlock_condition| unlock_condition.address())
             .unwrap()
     }
+
+    ///
```
Comment (Alex6323) [inline bee-message/src/output/basic.rs:178]: docs

File: bee-message/src/output/chain_id.rs [hunk 29]
```diff
@@ -11,14 +11,41 @@ pub enum ChainId {
...
+    pub fn is_null(&self) -> bool {
+        match self {
+            Self::Alias(alias_id) => alias_id.is_null(),
+            Self::Foundry(foundry_id) => foundry_id.is_null(),
+            Self::Nft(nft_id) => nft_id.is_null(),
+        }
+    }
+
+    ///
+    pub fn or_from_output_id(self, output_id: OutputId) -> ChainId {
```
Comment (Alex6323) [inline bee-message/src/output/chain_id.rs]: replace `ChainId` with `Self` in the return type

File: bee-message/src/output/chain_id.rs [hunk 30]
```diff
@@ -11,14 +11,41 @@ pub enum ChainId {
...
+}
+
 impl core::fmt::Display for ChainId {
     fn fmt(&self, f: &mut core::fmt::Formatter) -> core::fmt::Result {
         match self {
-            ChainId::Alias(alias) => write!(f, "{}", alias),
-            ChainId::Nft(nft) => write!(f, "{}", nft),
+            ChainId::Alias(id) => write!(f, "{}", id),
+            ChainId::Foundry(id) => write!(f, "{}", id),
+            ChainId::Nft(id) => write!(f, "{}", id),
```
Comment (Alex6323) [inline bee-message/src/output/chain_id.rs:48]: a little bit nit-picky, but reduces a bit of noise:
```rust
match self {
    ChaindId::Alias(id) => id,
    ....
}.fmt(f)
```

File: bee-message/src/output/foundry_id.rs [hunk 31]
```diff
@@ -15,19 +15,16 @@ string_serde_impl!(FoundryId);
...
-                .collect::<Vec<u8>>()
-                .try_into()
-                // SAFETY: the lengths are known.
-                .unwrap(),
-        )
+    pub fn build(alias_address: &AliasAddress, serial_number: u32, token_scheme: TokenScheme) -> Self {
+        let mut bytes = [0u8; FoundryId::LENGTH];
+        let mut packer = SlicePacker::new(&mut bytes);
+
+        // SAFETY: packing to an array of the correct length can't fail.
```
Comment (Alex6323) [inline bee-message/src/output/foundry_id.rs]: Might come across as too nit-picky, but the methods below aren't unsafe, and I think SAFETY tag is only used specifically for calls to unsafe code (although I am not 100% sure).

File: bee-message/src/output/mod.rs [hunk 32]
```diff
@@ -159,10 +163,58 @@ impl Output {
...
-    /// Verify if a valid storage deposit was made. Each [`Output`] has to have an amount that covers its associated
-    /// byte cost, given by [`ByteCostConfig`]. If there is a
-    /// [`StorageDepositReturnUnlockCondition`](unlock_condition::StorageDepositReturnUnlockCondition), its amount
-    /// is also checked.
+    /// Returns the chain identifier of an [`Output`], if any.
+    pub fn chain_id(&self) -> Option<ChainId> {
+        match self {
+            #[cfg(feature = "cpt2")]
+            Self::SignatureLockedSingle(_) => None,
+            #[cfg(feature = "cpt2")]
```
Comment (grtlr) [inline bee-message/src/output/mod.rs]: These `cpt2` features are not there anymore.

File: bee-message/src/output/alias.rs [hunk 33]
```diff
@@ -296,6 +298,141 @@ impl AliasOutput {
...
+        }
+
+        Ok(())
+    }
+
+    fn transition(
+        current_state: &Self,
+        next_state: &Self,
+        context: &ValidationContext,
+    ) -> Result<(), StateTransitionError> {
```
Comment (grtlr) [inline bee-message/src/output/alias.rs:373]: I think it would be cool to split up these big transition functions into smaller sub-functions with descriptive names. But that doesn't have to be done in this PR.

File: bee-message/src/output/mod.rs [hunk 34]
```diff
@@ -159,10 +163,58 @@ impl Output {
...
+            Self::Treasury(_) => None,
+            Self::Basic(_) => None,
+            Self::Alias(output) => Some(output.chain_id()),
+            Self::Foundry(output) => Some(output.chain_id()),
+            Self::Nft(output) => Some(output.chain_id()),
+        }
+    }
+
+    ///
+    pub fn state_transition(
```
Comment (Alex6323) [inline bee-message/src/output/mod.rs]: Wouldn't `validate_state_transition` be a better and more consistent name. It sets the expectation straight what to expect of this method, and not erroneously convey that state gets mutated.

File: bee-message/src/output/mod.rs [hunk 35]
```diff
@@ -229,3 +281,12 @@ fn minimum_storage_deposit(config: &ByteCostConfig, address: &Address) -> u64 {
         .unwrap();
     Output::Basic(basic_output).byte_cost(config)
 }
+
+///
+pub fn inputs_commitment<'a>(inputs: impl Iterator<Item = &'a Output>) -> [u8; 32] {
```
Comment (Alex6323) [inline bee-message/src/output/mod.rs]: docs

Comment (Alex6323) [reply bee-message/src/output/mod.rs]: `create_inputs_commitment` ?

Also maybe the corresponding docs could explain that the method generates a Blacke2b256 hash over the inputs so one understands as the consumer of this API what the returned bytes are.

File: bee-message/src/address/mod.rs [hunk 36]
```diff
@@ -67,4 +73,63 @@ impl Address {
...
+                let Signature::Ed25519(signature) = unlock_block.signature();
+
+                if signature.is_valid(&context.essence_hash, ed25519_address).is_err() {
+                    return Err(ConflictReason::InvalidSignature);
+                }
+
+                context.unlocked_addresses.insert(*self);
+            }
+            (Address::Ed25519(_ed25519_address), UnlockBlock::Reference(_unlock_block)) => {
+                // TODO actually check that it was unlocked by the same signature.
```
Comment (grtlr) [inline bee-message/src/address/mod.rs:114]: Will this be a separate PR?

File: bee-message/src/address/mod.rs (hunk 18 above)
Comment (thibault-martinez) [reply bee-message/src/address/mod.rs:92]: As said in the weekly, I will have another PR for documentation.

File: bee-api/bee-rest-api/src/types/dtos.rs (hunk 2 above)
Comment (thibault-martinez) [reply bee-api/bee-rest-api/src/types/dtos.rs]: Think you were reviewing an old commit, this is gone already

File: bee-message/src/unlock_block/mod.rs (hunk 3 above)
Comment (thibault-martinez) [reply bee-message/src/unlock_block/mod.rs]: Yeah we don't

File: bee-message/src/address/ed25519.rs (hunk 4 above)
Comment (thibault-martinez) [reply bee-message/src/address/ed25519.rs]: No

File: bee-message/tests/treasury_transaction_payload.rs (hunk 5 above)
Comment (thibault-martinez) [reply bee-message/tests/treasury_transaction_payload.rs:21]: No, we use it L24

File: bee-message/tip.md (hunk 23 above)
Comment (thibault-martinez) [reply bee-message/tip.md]: Whoops

File: bee-message/src/output/alias_id.rs (hunk 19 above)
Comment (thibault-martinez) [reply bee-message/src/output/alias_id.rs]: Yeah but your AliasId can be null and if that's the case it should be computed from an OutputId. That's very similar to Option's `or` but I'd prefer this one to be more explicit.

File: bee-message/src/output/alias.rs (hunk 24 above)
Comment (thibault-martinez) [reply bee-message/src/output/alias.rs]: Right

File: bee-message/src/output/alias.rs (hunk 25 above)
Comment (thibault-martinez) [reply bee-message/src/output/alias.rs]: Sure

File: bee-message/src/output/mod.rs (hunk 35 above)
Comment (thibault-martinez) [reply bee-message/src/output/mod.rs]: Honestly I don't really like that kind of prefixes, it's very much inconsistent as the boundary when to put it or not is not obvious at all. If we follow the logic, anything that returns ownership creates something so we should put it in a LOT of places in the repository.

File: bee-ledger/src/workers/consensus/white_flag.rs (hunk 15 above)
Comment (thibault-martinez) [reply bee-ledger/src/workers/consensus/white_flag.rs:72]: I'll think about an alternative but the inputs need to be indexable. With just an iterator we would have to do a lot of cloning of the iterator + a lot of iterations to index with `.nth`

File: bee-message/src/output/mod.rs (hunk 35 above)
Comment (Alex6323) [reply bee-message/src/output/mod.rs]: I don't know, but since methods/functions usually describe some kind of operation/action, I think in general it is good practice to prefix it with a verb. There are exceptions to that, most notably getters or  `len()`, `capacity()`-like methods in impl bocks. That's not the case here. It's a free function that is producing a hash from an input, so I think a verb prefix is justified.

You're call. Just wanted to give that input.

File: bee-ledger/src/workers/consensus/white_flag.rs [hunk 37]
```diff
@@ -68,54 +58,190 @@ fn apply_regular_essence<B: StorageBackend>(
...
+                output.unlock_conditions(),
+            ),
+            _ => return Err(Error::UnsupportedOutputKind(consumed_output.kind())),
+        };
+
+        if let Err(conflict) = conflict {
+            return Ok(conflict);
+        }
+
+        // TODO maybe put in unlocks methods.
```
Comment (Alex6323) [inline bee-ledger/src/workers/consensus/white_flag.rs:122]: Is that something you want to address later?

File: bee-ledger/src/workers/consensus/white_flag.rs [hunk 38]
```diff
@@ -68,54 +58,190 @@ fn apply_regular_essence<B: StorageBackend>(
...
-                // TODO
+    let inputs: Vec<(OutputId, &Output)> = consumed_outputs
+        .iter()
+        .map(|(output_id, created_output)| (*output_id, created_output.inner()))
+        .collect();
+
+    let mut context = ValidationContext::new(
+        transaction_id,
+        essence,
+        inputs.iter().map(|(output_id, input)| (output_id, *input)),
```
Comment (Alex6323) [inline bee-ledger/src/workers/consensus/white_flag.rs:77]: that's essentially copying the inputs? Why not create the `inputs` Vec with owned `OutputId` and `Output`. You deref the `input` here anyways.

File: bee-ledger/src/workers/consensus/white_flag.rs (hunk 15 above)
Comment (grtlr) [reply bee-ledger/src/workers/consensus/white_flag.rs:72]: Yeah, I agree that's also not ideal. I guess we can keep it as it is 👍

File: bee-ledger/src/workers/consensus/white_flag.rs [hunk 39]
```diff
@@ -68,54 +58,190 @@ fn apply_regular_essence<B: StorageBackend>(
...
+                output.amount(),
+                output.native_tokens(),
+                output.unlock_conditions(),
+            ),
+            Output::Nft(output) => (
+                output.unlock(output_id, unlock_block, &inputs, &mut context),
+                output.amount(),
+                output.native_tokens(),
+                output.unlock_conditions(),
+            ),
```
Comment (Alex6323) [inline bee-ledger/src/workers/consensus/white_flag.rs:114]: You could introduce an `OutputEssence` (or differently named) trait and implement it for all output types, so that the compiler knows that `unlock(..)`, `amount()`, etc are available for each of those. Then - I think - you can reduce the noise to:

```rust
use Output::*;
let ... = match consumed_output {
     Basic(output) | Alias(output) | Foundry(output) | Nft(output) => (
         <output as OutputEssence>.unlock(output_id, unlock_block, &inputs, &mut context),
         ...
    ),
   _ => ...
}
```

File: bee-ledger/src/workers/consensus/white_flag.rs (hunk 37 above)
Comment (thibault-martinez) [reply bee-ledger/src/workers/consensus/white_flag.rs:122]: Yes

File: bee-ledger/src/workers/consensus/white_flag.rs (hunk 38 above)
Comment (thibault-martinez) [reply bee-ledger/src/workers/consensus/white_flag.rs:77]: `Output` is not `Copy` so no it's not copying the inputs. We are creating an iterator over a Vec of `&Output` so the type ends up being `&&Output`, just removing one ref here.

File: bee-ledger/src/workers/consensus/white_flag.rs (hunk 39 above)
Comment (thibault-martinez) [reply bee-ledger/src/workers/consensus/white_flag.rs:114]: All the bindings in the same arm must have the same type so this kind of match is not legal

File: bee-message/src/output/mod.rs (hunk 34 above)
Comment (thibault-martinez) [reply bee-message/src/output/mod.rs]: I'll call it `verify_state_transition` to be consistent with our other verification functions

File: bee-message/src/output/foundry_id.rs (hunk 31 above)
Comment (thibault-martinez) [reply bee-message/src/output/foundry_id.rs]: I changed them to PANIC and opened an issue to also do it for the whole repo https://github.com/iotaledger/bee/issues/1270

File: bee-message/src/output/chain_id.rs (hunk 30 above)
Comment (thibault-martinez) [reply bee-message/src/output/chain_id.rs:48]: All the match arms must return the same type so this is not legal Rust

File: bee-message/src/output/alias.rs (hunk 33 above)
Comment (thibault-martinez) [reply bee-message/src/output/alias.rs:373]: I'd prefer to keep it as one logical function as this is a trait but what I can do is put more comments in it to separate and explain the checks.

File: bee-message/src/output/unlock_condition/immutable_alias_address.rs (hunk 17 above)
Comment (thibault-martinez) [reply bee-message/src/output/unlock_condition/immutable_alias_address.rs]: Done

File: bee-message/src/output/alias.rs (hunk 33 above)
Comment (thibault-martinez) [reply bee-message/src/output/alias.rs:373]: I'll do that when the TIPs are fully stable and I'll copy paste the checks at the relevant places.

File: bee-message/src/output/mod.rs (hunk 35 above)
Comment (thibault-martinez) [reply bee-message/src/output/mod.rs]: I'm not convinced as I still see it as being very much inconsistent but I'll do it.

File: bee-message/src/output/foundry.rs (hunk 26 above)
Comment (thibault-martinez) [reply bee-message/src/output/foundry.rs:362]: Yeah I need to think a bit more about it. Created an issue to not forget https://github.com/iotaledger/bee/issues/1272

File: bee-message/src/address/mod.rs (hunk 36 above)
Comment (thibault-martinez) [reply bee-message/src/address/mod.rs:114]: Same https://github.com/iotaledger/bee/issues/1272

File: bee-message/src/lib.rs (hunk 20 above)
Comment (thibault-martinez) [reply bee-message/src/lib.rs:35]: I would like to keep it like it if you don't mind.

File: bee-message/src/output/alias.rs (hunk 27 above)
Comment (thibault-martinez) [reply bee-message/src/output/alias.rs:374]: I can add `Verifier` to the trait name but I wouldn't add it to the method name. If I add it for transition I would need to add it to creation and destruction and that would be very redundant (there are even clippy checks that suggest you don't do that kind of things).

**Review by samuel-rufi — APPROVED** [review]

**Review by grtlr — APPROVED** [review]
